service/history/replication/executable_task.go

```diff
@@ -445,6 +461,166 @@ func (e *ExecutableTaskImpl) Resend(
+	)
+	if err != nil {
+		return err
+	}
+
+	engine, err := shardContext.GetEngine(ctx)
+	if err != nil {
+		return err
+	}
+
+	var eventsBatch [][]*historypb.HistoryEvent
+	var newRunEvents []*historypb.HistoryEvent
+	var versionHistory []*historyspb.VersionHistoryItem
+	const EmptyVersion = int64(-1) // 0 is a valid event version when namespace is local
+	var eventsVersion = EmptyVersion
+	isLastEvent := false
+	if len(newRunId) != 0 {
+		iterator := e.ProcessToolBox.RemoteHistoryFetcher.GetSingleWorkflowHistoryPaginatedIteratorInclusive(
+			ctx,
+			remoteCluster,
+			namespace.ID(workflowKey.NamespaceID),
+			workflowKey.WorkflowID,
+			newRunId,
+			1,
+			endEventVersion, // continue as new run's first event batch should have the same version as the last event of the old run
+			1,
+			endEventVersion,
+		)
+		iterator.HasNext()
+		batch, err := iterator.Next()
+		if err != nil {
+			return serviceerror.NewInternal(fmt.Sprintf("failed to get new run history when backfill: %v", err))
+		}
+		events, err := e.EventSerializer.DeserializeEvents(batch.RawEventBatch)
+		if err != nil {
+			return serviceerror.NewInternal(fmt.Sprintf("failed to deserailize run history events when backfill: %v", err))
+		}
+		newRunEvents = events
+	}
+
+	applyFn := func() error {
+		backFillRequest := &shard.BackfillHistoryEventsRequest{
+			WorkflowKey:         workflowKey,
+			SourceClusterName:   e.SourceClusterName(),
+			VersionedHistory:    e.ReplicationTask().VersionedTransition,
+			VersionHistoryItems: versionHistory,
+			Events:              eventsBatch,
+		}
+		if isLastEvent {
+			backFillRequest.NewEvents = newRunEvents
+			backFillRequest.NewRunID = newRunId
+		}
+		err := engine.BackfillHistoryEvents(ctx, backFillRequest)
+		if err != nil {
+			return serviceerror.NewInternal(fmt.Sprintf("failed to backfill: %v", err))
+		}
+		eventsBatch = nil
+		versionHistory = nil
+		eventsVersion = EmptyVersion
+		return nil
+	}
+	iterator := e.ProcessToolBox.RemoteHistoryFetcher.GetSingleWorkflowHistoryPaginatedIteratorInclusive(
+		ctx,
+		remoteCluster,
+		namespace.ID(workflowKey.NamespaceID),
+		workflowKey.WorkflowID,
+		workflowKey.RunID,
+		startEventId,
+		startEventVersion,
+		endEventId,
+		endEventVersion,
+	)
+	for iterator.HasNext() {
+		batch, err := iterator.Next()
+		if err != nil {
+			return err
+		}
+		events, err := e.EventSerializer.DeserializeEvents(batch.RawEventBatch)
+		if err != nil {
+			return err
+		}
+		if len(events) == 0 {
+			return serviceerror.NewInvalidArgument("Empty batch received from remote during resend")
+		}
+		if len(eventsBatch) != 0 && len(versionHistory) != 0 {
+			if !versionhistory.IsEqualVersionHistoryItems(versionHistory, batch.VersionHistory.Items) ||
+				(eventsVersion != EmptyVersion && eventsVersion != events[0].Version) {
+				err := applyFn()
+				if err != nil {
+					return err
+				}
+			}
+		}
+		eventsBatch = append(eventsBatch, events)
+		if events[len(events)-1].GetEventId() == endEventId {
```

Comment:
Is it possible that endEventId falling into the middle of a batch?

Reply:
new run logic handling here is only supposed to be used by verify versioned transition task where no events are included in the tasks. So only the verify tasks that has new runId is possible to exercise this code path. I can add additional validation to make sure it is verify tasks that is providing newRunId.